addons/block_code/blocks/communication/groups.gd

```diff
@@ -0,0 +1,70 @@
+@tool
+## Common BlockExtension for scene group options.
+extends BlockExtension
+
+const OptionData = preload("res://addons/block_code/code_generation/option_data.gd")
+const Util = preload("res://addons/block_code/ui/util.gd")
+
+
+# Global groups are just project settings in the global_group group.
+# ProjectSettings doesn't offer an API to get the project groups or to get all
+# settings. Fortunately, settings are simply implemented as properties on the
+# ProjectSettings object.
+static func _add_global_groups(groups: Dictionary):
+	for prop in ProjectSettings.get_property_list():
+		var name = prop["name"]
+		if name.begins_with("global_group/"):
+			var parts = name.split("/", false, 1)
+			groups[parts[1]] = null
+
+
+# Add all the groups in a node and its children, ignoring internal _ prefixed
+# groups.
+static func _add_node_groups(groups: Dictionary, node: Node, scene_root: Node):
+	# Ignore this node and its children if it isn't part of the scene.
+	if node != scene_root and node.owner != scene_root:
```

Comment:
One issue with this use of `node.owner` is we're going to skip nodes in an instantiated scene. For instance, with our pong_game.tscn example, the BlockCode for Ball refers to a group named "walls", but that group isn't in the generated options list because we created that group in the separate Space scene. Was this intentional? Otherwise, it looks to me what we want here is more like...

```suggestion
	if node != scene_root and not scene_root.is_ancestor_of(node):
```

Reply:
That was intentional. Since that's a different scene, it didn't seem appropriate to include it's groups since that's not how Godot itself handles them. Hmm, actually I guess that's not true testing it out. After adding a group in a child scene, the Groups Editor shows it as available in the parent scene but with a :lock: and a tooltip saying "This group belongs to another scene and can't be edited".

If we're just going to descend all the child scenes then we can just drop this conditional completely. We already know it's an ancestor since the function is walking children recursively. Lemme try that out...